src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java

```diff
@@ -0,0 +1,151 @@
+    protected void setMigrationCriteria() {
+        // No migration criteria currently needed.
+    }
+
+    @Override
+    protected boolean isMigrationNeeded(Course entity) {
+        return true;
+    }
+
+    @Override
+    protected void migrateEntity(Course oldCourse) throws Exception {
+        teammates.storage.sqlentity.Course newCourse = createCourse(oldCourse);
+        // TODO: add shutdown hook to save the entity
+        // Runnable shutdownScript = () -> { cascadeDelete(newCourse)};
+        // Runtime.getRuntime().addShutdownHook(new Thread(shutdownScript));
+
+        migrateCourseEntity(newCourse);
+        // verifyCourseEntity(newCourse);
+        // markOldCourseAsMigrated(courseId)
+        // Runtime.getRuntime().removeShutDownHook(new Thread(shutdownScript));
+    }
+
+    private void migrateCourseEntity(teammates.storage.sqlentity.Course newCourse) {
+        Map<String, teammates.storage.sqlentity.Section> sectionNameToSectionMap = migrateSectionChain(newCourse);
+        System.out.println(sectionNameToSectionMap); // To stop lint from complaining
+        // TODO: Add mirgrateFeedbackChain
+        // migrateFeedbackChain(sectionNameToSectionMap);
+    }
+
+    private Map<String, teammates.storage.sqlentity.Section> migrateSectionChain(
+            teammates.storage.sqlentity.Course newCourse) {
+        List<CourseStudent> oldStudents = ofy().load().type(CourseStudent.class).filter("courseId", newCourse.getId())
+                .list();
+        Map<String, teammates.storage.sqlentity.Section> sections = new HashMap<>();
+        Map<String, List<CourseStudent>> sectionToStuMap = oldStudents.stream()
+                .collect(Collectors.groupingBy(CourseStudent::getSectionName));
+
+        for (Map.Entry<String, List<CourseStudent>> entry : sectionToStuMap.entrySet()) {
+            String sectionName = entry.getKey();
+            List<CourseStudent> stuList = entry.getValue();
+            teammates.storage.sqlentity.Section newSection = createSection(newCourse, sectionName);
+            sections.put(sectionName, newSection);
+            migrateTeams(newCourse, newSection, stuList);
+            saveEntityDeferred(newSection);
+        }
+        return sections;
+    }
+
+    private void migrateTeams(teammates.storage.sqlentity.Course newCourse,
+            teammates.storage.sqlentity.Section newSection, List<CourseStudent> studentsInSection) {
+        Map<String, List<CourseStudent>> teamNameToStuMap = studentsInSection.stream()
+                .collect(Collectors.groupingBy(CourseStudent::getTeamName));
+        for (Map.Entry<String, List<CourseStudent>> entry : teamNameToStuMap.entrySet()) {
+            String teamName = entry.getKey();
+            List<CourseStudent> stuList = entry.getValue();
+            teammates.storage.sqlentity.Team newTeam = createTeam(newSection, teamName);
+            migrateStudents(newCourse, newTeam, stuList);
+            saveEntityDeferred(newTeam);
```

Comment:
Just wondering, if we create the team after the student, would there be a case Hibernate gives an exception where we try to create a Student with a non-existent team? (If by coincidence they are flushed separately, following the new design).

Reply:
This is a good catch!